commit e875132d921c1c522ddbe0a3213f0b6110bacc5b
Author: Zhaofeng Li <hello@zhaofeng.li>
Date:   Sat Jan 15 16:06:37 2022 -0800

    gallivm: Fix target CPU spec on RISC-V

diff --git a/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp b/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
index be288ab02e2..401e089e840 100644
--- a/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
+++ b/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
@@ -535,6 +535,22 @@ lp_build_create_jit_compiler_for_module(LLVMExecutionEngineRef *OutJIT,
       MCPU = util_get_cpu_caps()->has_msa ? "mips64r5" : "mips64r2";
 #endif
 
+#ifdef PIPE_ARCH_RISCV
+   /*
+    * The RISC-V target in LLVM doesn't have a "generic" CPU, only "generic-rv32"
+    * and "generic-rv64".
+    *
+    * This was fixed in LLVM master in <https://reviews.llvm.org/D105274>.
+    */
+#ifdef PIPE_ARCH_RISCV64
+   if (MCPU == "generic")
+       MCPU = "generic-rv64";
+#else
+   if (MCPU == "generic")
+       MCPU = "generic-rv32";
+#endif
+#endif
+
    builder.setMCPU(MCPU);
    if (gallivm_debug & (GALLIVM_DEBUG_IR | GALLIVM_DEBUG_ASM | GALLIVM_DEBUG_DUMP_BC)) {
       debug_printf("llc -mcpu option: %s\n", MCPU.str().c_str());
diff --git a/src/gallium/include/pipe/p_config.h b/src/gallium/include/pipe/p_config.h
index 978aa455ecb..cb8e2ea6627 100644
--- a/src/gallium/include/pipe/p_config.h
+++ b/src/gallium/include/pipe/p_config.h
@@ -130,6 +130,19 @@
 #define  PIPE_ARCH_MIPS
 #endif
 
+#if defined(__riscv) || defined(__riscv__)
+#define PIPE_ARCH_RISCV
+
+#if __riscv_xlen == 64
+#define PIPE_ARCH_RISCV64
+#else
+#error XXX - Not 64-bit?
+#endif
+
+#else
+#error XXX - Not RISC-V?
+#endif
+
 /*
  * Endian detection.
  */
